### ISSUE TITLE
Guidance for best practice around exposing consul ports for servers

### ISSUE THREAD

**Issue by Throckmortra** [hashicorp/consul-k8s#942]
#### Question

hello,

I'm attempting to run some Consul servers on EKS Kubernetes but have run into some uncertainty on how best to expose the [required server ports](https://www.consul.io/docs/install/ports) such as serflan/wan. Most of my issues have been around [mixing protocols in Kubernetes loadbalancers](https://learn.hashicorp.com/tutorials/consul/kubernetes-eks-aws?utm_source=consul.io&utm_medium=docs&utm_content=k8s&utm_term=eks) but I ran across [this post](https://discuss.hashicorp.com/t/consul-udp-headaches-on-all-public-clouds/26703/4) on the Hashicorp forums that gave me pause. Is it the case that we should not be loadbalancing consul gossip ports? And if so, what is best practice for this? For my purposes, I'm guessing I could rig something up with EIPs on my nodes but Kubernetes is ephemeral architecture and this approach feels strange as well. 

[This official guide ](https://learn.hashicorp.com/tutorials/consul/kubernetes-eks-aws?utm_source=consul.io&utm_medium=docs&utm_content=k8s&utm_term=eks) and the [reference architecture](https://learn.hashicorp.com/tutorials/consul/kubernetes-reference-architecture) do not seem to mention this.

TL;DR: What do you recommend for permanently exposing all the required consul server ports for service mesh/gossip on ephemeral Kubernetes clusters?

Thanks!

**Comment by Throckmortra** [comment]
after spending some more time on this, I believe I have found how they want you to do this.
Some notes:

- it is the case that you should not be load balancing consul servers, they must be directly addressable within the LAN
- in the free-version of consul, they offer [Mesh Gateways](https://www.consul.io/docs/connect/gateways#mesh-gateways) for Datacenter-to-datacenter communication to get around the requirement of addressing nodes directly
- if you have a network topology like mine (many kubernetes clusters, many cloud providers), you need to pay for an Enterprise license to unlock the [Admin Partition feature](https://www.consul.io/docs/connect/gateways/mesh-gateway/service-to-service-traffic-partitions)
- Admin partitions allow you to use the mesh gateway to address consul nodes/servers in a "LAN" where node-to-node communication would not normally be possible over a regional/local network 

Hope this helps some of you

**Comment by ishustava** [comment]
Hey @Throckmortra 

Thanks for posting this summary. Yes, that would be our recommendation. Here are some more docs around how to use WAN federation via mesh gateways on k8s: https://www.consul.io/docs/k8s/installation/multi-cluster

I'm going to go ahead and close this issue as it seems that your question is answered, but please let me know if that is not the case!

**Comment by Throckmortra** [comment]
What performance or scaling issues could we encounter if we attempted to make thousands of WANs?

**Comment by ishustava** [comment]
At the moment, we don't have any published data around performance of the multi-cluster service mesh with WAN federation. I don't think I can have a good answer here since this is quite a broad question, but here are few things to keep in mind:

1. Mesh gateway that is used for cross-cluster communication is another envoy proxy. We have some information about how consul performs within a single cluster that could help a bit published in the [Consul Global Scale Benchmark](https://www.hashicorp.com/cgsb).
2. If you put your mesh gateways behind a load balancer, you have to factor the performance of that in.
3. Mesh gateways are used for dc-to-dc WAN gossip, and there quite a few ways to tune those parameters described in [these docs](https://www.consul.io/docs/install/performance).
4. You need to factor in things like cluster size, the types of workloads you have to understand the type of traffic that will be going through the mesh gateways.

Since this is a different topic from the original issue, I'd recommend opening a new issue if you think that this is something you want us to track separately.
